### ISSUE TITLE
Pulumi should be able to wait until the current operation has finished before applying

### ISSUE THREAD

**Issue by squarebracket** [pulumi/pulumi#4052]
It would be great if, instead of simply bailing out when an update is in progress, it could wait until the current update has finished before applying the requested update. This would be of great benefit when it comes to running pulumi in CI. Something like a `pulumi wait [-s stack]` command would be amazing.

Currently, I'm using some bash to "fix" the problem, by `curl`ing the private API used by the console, and waiting if the stack has a `currentOperation` property. However, this has some problems, namely that there is non-zero time between making the `curl` and applying the update, so the operation can still fail.

Some possible nice-to-haves:
1. a `--timeout` flag so that it will still bail after waiting for a certain amount of time
2. queues, so that updates are processed in the order they are first requested

**Comment by leezen** [comment]
Could you please describe this scenario in a bit more detail? Is the scenario something like you've pushed a commit, CI runs `pulumi up` and while that's happening, you push another commit, which causes another job to kick off, which attempts to run a new `pulumi up` in parallel? Just want to make sure I understand the use-case you have in mind.

**Comment by KaoruDev** [comment]
I can't speak for squarebracket's scenario, but @leezen what you described is what I'm trying to figure out how to solve. Essentially, there are scenarios where pulumi is running an update to `commit-1`, is taking a long time because there are a lot of resources to create, and the CI will start running `commit-2` because it was merged right after.

It would be great if there was a way to "queue" these updates via Pulumi.

**Comment by Frassle** [comment]
Pulumi already understand if stacks are locked because another deployment is running. Could you implement queues by just sleeping and retrying if you get the stack locked error?

**Comment by Frassle** [comment]
I'm not keen on adding `pulumi wait` that sounds like a recipe for race bugs, but I think we could add a `--timeout` to up to keep trying to take the lock.
